expression/generator/compare_vec.go

```diff
@@ -0,0 +1,316 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// +build ignore
+
+package main
+
+import (
+	"bytes"
+	"flag"
+	"go/format"
+	"io/ioutil"
+	"log"
+	"path/filepath"
+	"text/template"
+
+	. "github.com/pingcap/tidb/expression/generator/helper"
+)
+
+const header = `// Copyright 2019 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Code generated by go generate in expression/generator; DO NOT EDIT.
+
+package expression
+`
+
+const newLine = "\n"
+
+const builtinCompareImports = `import (
+	"github.com/pingcap/tidb/types"
+	"github.com/pingcap/tidb/types/json"
+	"github.com/pingcap/tidb/util/chunk"
+)
+`
+
+var builtinCompareVecTpl = template.Must(template.New("").Parse(`
+func (b *builtin{{ .compare.CompareName }}{{ .type.TypeName }}Sig) vecEvalInt(input *chunk.Chunk, result *chunk.Column) error {
+	n := input.NumRows()
+	buf0, err := b.bufAllocator.get(types.ET{{ .type.ETName }}, n)
+	if err != nil {
+		return err
+	}
+	defer b.bufAllocator.put(buf0)
+	if err := b.args[0].VecEval{{ .type.TypeName }}(b.ctx, input, buf0); err != nil {
+		return err
+	}
+	buf1, err := b.bufAllocator.get(types.ET{{ .type.ETName }}, n)
+	if err != nil {
+		return err
+	}
+	defer b.bufAllocator.put(buf1)
+	if err := b.args[1].VecEval{{ .type.TypeName }}(b.ctx, input, buf1); err != nil {
+		return err
+	}
+
+{{ if .type.Fixed }}
+	arg0 := buf0.{{ .type.TypeNameInColumn }}s()
+	arg1 := buf1.{{ .type.TypeNameInColumn }}s()
+{{ end }}
+	result.ResizeInt64(n, false)
+	result.MergeNulls(buf0, buf1)
+	i64s := result.Int64s()
+	for i := 0; i < n; i++ {
+		if result.IsNull(i) {
+			continue
+		}{{ if eq .type.ETName "Json" }}
+		val := json.CompareBinary(buf0.GetJSON(i), buf1.GetJSON(i)){{ else if eq .type.ETName "Real" }}
+		val := types.CompareFloat64(arg0[i], arg1[i]){{ else if eq .type.ETName "String" }}
+		val := types.CompareString(buf0.GetString(i), buf1.GetString(i)){{ else if eq .type.ETName "Duration" }}
+		val := types.CompareDuration(arg0[i], arg1[i]){{ else if eq .type.ETName "Datetime" }}
+		val := arg0[i].Compare(arg1[i]){{ else if eq .type.ETName "Decimal" }}
+		val := arg0[i].Compare(&arg1[i]){{ end }}
```

Comment:
nit: It's acceptable to have empty lines in generated files, IMO, it'd be better format `if else` here to make generator code more readable.

Reply:
ok